### PR TITLE
 resolve Check whether special is active by tag name or by id (bis)

### PR DESCRIPTION
**Pull request by newfrenchy83**
replace https://github.com/wesnoth/wesnoth/pull/3920 PR for resolve https://github.com/wesnoth/wesnoth/issues/3915

### CODE REVIEW

**Comment by CelticMinstrel** [comment]
Just noting that I don't see any problems with this; it _probably_ can be merged.

File: src/units/attack_type.hpp [hunk 1]
```diff
@@ -73,7 +73,7 @@ class attack_type : public std::enable_shared_from_this<attack_type>
 
 	// In unit_abilities.cpp:
 
-	bool get_special_bool(const std::string& special, bool simple_check=false) const;
+	bool get_special_bool(const std::string& special, bool simple_check=false, bool special_id=true, bool special_tags=true) const;
```
Comment (jostephd) [inline src/units/attack_type.hpp]: Please document the two new parameters:

```
/// @return True iff the special @a special is active.
/// @param special_id If true, match @a special against the @c id of special tags.
/// @param special_tags If true, match @a special against the tag name of special tags.
```

File: src/units/attack_type.cpp [hunk 2]
```diff
@@ -135,7 +137,7 @@ static bool matches_simple_filter(const attack_type & attack, const config & fil
 	if(!filter_special.empty()) {
 		bool found = false;
 		for(auto& special : filter_special) {
-			if(attack.get_special_bool(special, true)) {
+			if(attack.get_special_bool(special, true, true, false)) {
```
Comment (jostephd) [inline src/units/attack_type.cpp]: With this change, the meaning of existing WML code that uses filter_weapon with `special="foo"` will be changed. The language in https://wiki.wesnoth.org/FilterWML#Filtering_Weapons is unclear, but it seems reasonable that there's UMC out there that relies on `special=foo` matching either tag name or id. So, would it be preferable to keep the meaning of `special=foo` as-is, deprecate it, and add _two_ new keys, `special_id=foo` and `special_tag=foo` (as well as their `special_*_active` variants) instead?

Comment (newfrenchy83) [reply src/units/attack_type.cpp]: i will doing that then.

Comment (jostephd) [reply src/units/attack_type.cpp]: Thanks! If you have any question, feel free to stop by the development channel on discord/IRC. See links here: https://github.com/wesnoth/wesnoth/blob/master/.github/CONTRIBUTING.md#contacting-us

Comment (newfrenchy83) [reply src/units/attack_type.cpp]: it is done

**Comment by newfrenchy83** [comment]
i have put a test case in https://github.com/wesnoth/wesnoth/pull/3920/files but the filter_weapon in drains abilities is in wrong place but i must wait what https://github.com/wesnoth/wesnoth/pull/4435 will be merged before to put a test case with true code in this PR.

File: src/units/attack_type.cpp (hunk 2 above)
Comment (jostephd) [reply src/units/attack_type.cpp]: Thanks! What's your opinion regarding whether or not to deprecate `"special"` and `"special_active"`?

Comment (newfrenchy83) [reply src/units/attack_type.cpp]: Well we can mark "special" as deprecate but keep it for version 1.16 as the beta version is imminent now and delete it only for the following versions to not take the UMC devellopeurs to the devoid

Comment (jostephd) [reply src/units/attack_type.cpp]: > but keep it for version 1.16

Of course! That's the project policy for deprecations already: https://wiki.wesnoth.org/CompatibilityStandards

Okay, then let's deprecate both `special` and `special_active`. To do that, call [deprecated_message](https://github.com/wesnoth/wesnoth/blob/a0ccc2dff42996c3e2a0ba4f3fb19f259540dbe5/src/deprecation.hpp#L21-L44) if either of those keys is set. I suggest to deprecate at DEP_LEVEL::PREEMPTIVE.

**Comment by jostephd** [comment]
> i have put a test case in https://github.com/wesnoth/wesnoth/pull/3920/files but the filter_weapon in drains abilities is in wrong place but i must wait what #4435 will be merged before to put a test case with true code in this PR.

Thanks. I understand that #4435 blocks adding that test case to this PR, but couldn't we merge this PR before #4435? I realize the test case in #3920 wouldn't work yet, but we could test the feature differently to be sure it works. For example, we could use a StandardUnitFilter with a [has_attack] subtag.

File: src/units/attack_type.cpp [hunk 3]
```diff
@@ -137,6 +137,7 @@ static bool matches_simple_filter(const attack_type & attack, const config & fil
 		return false;
 
 	if(!filter_special.empty()) {
+		deprecated_message("special=", DEP_LEVEL::PREEMPTIVE, {1, 16, 0}, "Please use special_id or special_type instead");
```
Comment (jostephd) [inline src/units/attack_type.cpp]: Looks good, but the version should be `{1, 17, 0}`.

File: src/units/attack_type.cpp [hunk 4]
```diff
@@ -164,6 +165,7 @@ static bool matches_simple_filter(const attack_type & attack, const config & fil
 
 
 	if(!filter_special_active.empty()) {
+		deprecated_message("special_active=", DEP_LEVEL::PREEMPTIVE, {1, 16, 0}, "Please use special_id_active or special_type_active instead");
```
Comment (jostephd) [inline src/units/attack_type.cpp]: Looks good, but the version should be `{1, 17, 0}`.

Comment (newfrenchy83) [reply src/units/attack_type.cpp]: it is done

File: src/units/attack_type.cpp (hunk 2 above)
Comment (jostephd) [reply src/units/attack_type.cpp]: Don't forget to include deprecation.hpp

File: src/units/attack_type.cpp (hunk 4 above)
Comment (jostephd) [reply src/units/attack_type.cpp]: You need to include deprecation.hpp and game_version.hpp to fix a build error.

**Comment by jostephd** [comment]
@newfrenchy83 With this PR, I started HttT S6 _The Siege of Elensefar_ in debug mode and ran the following commands in the lua console:

```
$ wesnoth.get_units { wml.tag.has_attack { special = 'magical' }  }
{[1]=unit: 0x5587827722d8}
$ wesnoth.get_units { wml.tag.has_attack { special_id = 'magical' }  }
{}
$ wesnoth.get_units { wml.tag.has_attack { special_type = 'chance_to_hit' }  }
{[1]=unit: 0x55879f520158}
```

Why did the second one return nothing?

**Comment by newfrenchy83** [comment]
@jostephd i fixed special_id and corrected the test case.

**Comment by newfrenchy83** [comment]
- 

> 
> 
> @newfrenchy83 With this PR, I started HttT S6 _The Siege of Elensefar_ in debug mode and ran the following commands in the lua console:
> 
> ```
> $ wesnoth.get_units { wml.tag.has_attack { special = 'magical' }  }
> {[1]=unit: 0x5587827722d8}
> $ wesnoth.get_units { wml.tag.has_attack { special_id = 'magical' }  }
> {}
> $ wesnoth.get_units { wml.tag.has_attack { special_type = 'chance_to_hit' }  }
> {[1]=unit: 0x55879f520158}
> ```
> 
> Why did the second one return nothing?



> 
> 
> @newfrenchy83 With this PR, I started HttT S6 _The Siege of Elensefar_ in debug mode and ran the following commands in the lua console:
> 
> ```
> $ wesnoth.get_units { wml.tag.has_attack { special = 'magical' }  }
> {[1]=unit: 0x5587827722d8}
> $ wesnoth.get_units { wml.tag.has_attack { special_id = 'magical' }  }
> {}
> $ wesnoth.get_units { wml.tag.has_attack { special_type = 'chance_to_hit' }  }
> {[1]=unit: 0x55879f520158}
> ```
> 
> Why did the second one return nothing?

I haven't changed special to special_id in code, this is fixed now

File: src/units/attack_type.hpp (hunk 1 above)
Comment (jostephd) [reply src/units/attack_type.hpp]: @newfrenchy83 Ping

File: data/scenario-test.cfg [hunk 5]
```diff
@@ -532,6 +532,71 @@ Xu, Xu, Qxu, Qxu, Ql, Ql, Ql, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Gg, Gg, Gg, Gg, Gg
...
+        name=attack_end
+        first_time_only=yes
+
+        [filter_attack]
+            special=chance_to_hit
+        [/filter_attack]
+
+        [message]
+            speaker=unit
+            message="I used a chance_to_hit attack specials."
```
Comment (jostephd) [inline data/scenario-test.cfg]: Change `specials` to `special`.

File: data/scenario-test.cfg [hunk 6]
```diff
@@ -532,6 +532,71 @@ Xu, Xu, Qxu, Qxu, Ql, Ql, Ql, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Gg, Gg, Gg, Gg, Gg
...
+        name=attack_end
+        first_time_only=yes
+
+        [filter_attack]
+            special_id=chance_to_hit
+        [/filter_attack]
+
+        [message]
+            speaker=unit
+            message="a weapon specials with id=chance_to_hit, serious?"
```
Comment (jostephd) [inline data/scenario-test.cfg]: Please make the message clearer:

`message="BUG: This message should never be displayed!"`

File: data/scenario-test.cfg [hunk 7]
```diff
@@ -532,6 +532,71 @@ Xu, Xu, Qxu, Qxu, Ql, Ql, Ql, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Gg, Gg, Gg, Gg, Gg
...
+        name=attack_end
+        first_time_only=yes
+
+        [filter_attack]
+            special_type=magical
+        [/filter_attack]
+
+        [message]
+            speaker=unit
+            message="a magical special attack type, serious?"
```
Comment (jostephd) [inline data/scenario-test.cfg]: Please make the message clearer:

`message="BUG: This message should never be displayed!"`

File: data/scenario-test.cfg [hunk 8]
```diff
@@ -532,6 +532,71 @@ Xu, Xu, Qxu, Qxu, Ql, Ql, Ql, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Xu, Gg, Gg, Gg, Gg, Gg
...
+
+        [filter_attack]
+            special=chance_to_hit
+        [/filter_attack]
+
+        [message]
+            speaker=unit
+            message="I used a chance_to_hit attack specials."
+        [/message]
+    [/event]
```
Comment (jostephd) [inline data/scenario-test.cfg]: I notice there's no `[filter_attack] special=magical` case. Is that intentional?

Comment (newfrenchy83) [reply data/scenario-test.cfg]: Oops i forget this, i will correct test. No another change plannef

**Comment by newfrenchy83** [comment]
@jostephd it is done you can doing final review pass

**Comment by jostephd** [comment]
@newfrenchy83 Please address https://github.com/wesnoth/wesnoth/pull/4424#discussion_r334253543. With that changed, I would be happy to merge this.

File: src/units/abilities.cpp [hunk 9]
```diff
@@ -667,15 +715,27 @@ bool attack_type::get_special_bool(const std::string& special, bool simple_check
...
 	std::vector<special_match> special_tag_matches;
 	std::vector<special_match> special_id_matches;
-	get_special_children(special_tag_matches, special_id_matches, other_attack_->specials_, special);
-	for(const special_match& entry : special_tag_matches) {
-		if ( other_attack_->special_active(*entry.cfg, AFFECT_OTHER, entry.tag_name) ) {
-			return true;
+	if(special_id && special_tags){
+		get_special_children(special_tag_matches, special_id_matches, other_attack_->specials_, special);
+	}
+	else if (special_id && !special_tags){
```
Comment (jostephd) [inline src/units/abilities.cpp]: The placement of `else if` is inconsistent here compared to earlier in the function. I won't ask you to fix it this time, but in the future please be consistent about placement of `else` and about (lack of) spaces after `if`.

**Comment by newfrenchy83** [comment]
@jostephd i put doxygen documentation in attack_type.cpp

**Comment by jostephd** [comment]
> @jostephd i put doxygen documentation in attack_type.cpp

I don't think that would work. I think doxygen only recognizes comments in function declarations and definitions. What's more, you deleted the `@` from `@param`, so doxygen won't recognize it anyway.

I'll fix it and merge.

**Comment by jostephd** [comment]
Merged in a87be483073.

@newfrenchy83 Please address the **first two items** of https://github.com/wesnoth/wesnoth/pull/4424#pullrequestreview-301086942. And thank you very much for the pull request and for being so patient with how long it took us to review and merge it! :+1:

**Comment by newfrenchy83** [comment]
> 
> 
> Merged in [a87be48](https://github.com/wesnoth/wesnoth/commit/a87be4830733c7bffeccdd01264afb66bf1d70f9).
> 
> @newfrenchy83 Please address the **first two items** of [#4424 (review)](https://github.com/wesnoth/wesnoth/pull/4424#pullrequestreview-301086942). And thank you very much for the pull request and for being so patient with how long it took us to review and merge it! +1

@jostephd i don't know how doing that

**Comment by jostephd** [comment]
@newfrenchy83 I went ahead and made the change myself, see https://wiki.wesnoth.org/index.php?title=FilterWML&diff=64960&oldid=60669. Please review that, is the documentation correct? Did I leave anything out?

**Comment by newfrenchy83** [comment]
it is correct for me

**Comment by jostephd** [comment]
Thanks for the review!

**Pull request closed without merging**